src/Stratis.Bitcoin.Features.Consensus/Behaviors/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -92,16 +111,107 @@ protected override Payload ConstructHeadersPayload(BlockLocator locator, uint256
+                this.consensusManager,
+                this.peerBanning,
+                this.loggerFactory,
+                this.network,
+                this.chainState);
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer supports Proven Headers and PH has been activated.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if is peer is PH enabled; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerPhEnabled(INetworkPeer peer)
+        {
+            return peer.Version >= NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION;
+        }
+
+        /// <summary>
+        /// Determines whether the specified peer is Whitelisted.
+        /// </summary>
+        /// <param name="peer">The peer.</param>
+        /// <returns>
+        ///   <c>true</c> if the specified peer is Whitelisted; otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsPeerWhitelisted(INetworkPeer peer)
+        {
+            return peer.Behavior<IConnectionManagerBehavior>()?.Whitelisted == true;
+        }
+
+
+        private bool IsProvenHeaderActivated()
+        {
+            if (this.network.Consensus.Options is PosConsensusOptions options)
+            {
+                long currentHeight = this.chainState.ConsensusTip.Height;
+                return (options.ProvenHeadersActivationHeight > 0) && (currentHeight >= options.ProvenHeadersActivationHeight);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Builds the GetHeadersPayload.
+        /// </summary>
+        /// <returns>The GetHeadersPayload instance.
+        /// If the peer can serve PH, GetProvenHeadersPayload is returned, otherwise if it's a legacy peer but it's whitelisted,
+        /// GetHeadersPayload is returned.
+        /// If the attached peer is a legacy peer and it's not whitelisted, returns null.
+        /// </returns>
+        protected override GetHeadersPayload BuildGetHeadersPayload()
+        {
+            INetworkPeer peer = this.AttachedPeer;
+
+            if (this.IsProvenHeaderActivated())
+            {
+                if (this.IsPeerPhEnabled(peer))
+                {
+                    return new GetProvenHeadersPayload()
+                    {
+                        BlockLocator = (this.ExpectedPeerTip ?? this.consensusManager.Tip).GetLocator(),
+                        HashStop = null
+                    };
+                }
+                // If the peer doesn't supports PH but it's whitelisted, issue a standard GetHeadersPayload
+                else if (IsPeerWhitelisted(peer) || this.IsProvenHeaderActivated())
+                    return base.BuildGetHeadersPayload();
+                // If the peer doesn't support PH and isn't whitelisted, return null (stop synch attempt with legacy StratisX nodes).
+                else
+                    return null;
```

Comment:
I still think we should return a normal GetHeaders if PH is not enabled and peer is not upgraded.

So we need to return `GetProvenHeadersPayload` if PH enabled and protocol >= PH protocol otherwise normal headers.

Reply:
The whitelistedness is for processing the headers not for getheaders so the other method is correct (i.e processes legacy) but getheaders is no big deal.
We can have a vid chant perhaps to talk about this when you have time.

Reply:
ok but AFAIU BuildGetHeadersPayload is used to create the GetHeaders payload sent by our node to the other, when we want to sync to a peer, so we are requiring nodes in the legacy way and this is what we don't want to: we want to synch in a legacy way only to whitelisted nodes or before ph activation.

Actually we serve GetHeaders requests (so HeadersPayload) to any node, so any node can ask us headers in a legacy way and we serve them, but not the opposite (we don't want to request legacy node to peers we don't trust or that don't serve PH headers)

Reply:
@noescape00 what's your thought about this

Reply:
No this is only about triggering the sync, the handling of headers is done somewhere else (bellow in the method `ProcessLegacyHeadersAsync`) there we should ignore regular headers if we are PH activated.

Reply:
`GetHeadersPayload` is handled only in ConsensusManagerBehavior.cs
https://github.com/stratisproject/StratisBitcoinFullNode/blob/e7fd207adf857a3f2ecc85261ba3ca3fe87c2846/src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs#L135

This mean that whenever we receive a `getheaders` message from a peer, we handle that message only in that method (ProvenHeadersConsensusManagerBehavior inherit from ConsensusManagerBehavior so it handle the same way) and what we actually do is serving the request, so we allow any peer to ask us legacy headers. Isn't this right?
What we prevent, is asking legacy headers from peers that aren't whitelisted (after ph is activated)

Basically you started saying:
>I still think we should return a normal GetHeaders if PH is not enabled and peer is not upgraded.

And we are already doing it, because when a peer ask us headers using `getheaders` message, we handle it normally as before, see
https://github.com/stratisproject/StratisBitcoinFullNode/blob/e7fd207adf857a3f2ecc85261ba3ca3fe87c2846/src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs#L135-L137

note that `ProcessGetHeadersAsync` method is called both when legacy `getheaders` or new  `getprovhdr` is received, what change is message we send to that peer as a reply to its request:
https://github.com/stratisproject/StratisBitcoinFullNode/blob/e7fd207adf857a3f2ecc85261ba3ca3fe87c2846/src/Stratis.Bitcoin/Consensus/ConsensusManagerBehavior.cs#L180

Now, taking a closer look to `ConstructHeadersPayload` method now i see another problem: that method actually always return a proven header payload because when overridden in `ConstructHeadersPayload` it doesn't check if the request was a legacy one or a PH one.
So this is what should be fixed (this wasn't part of my changes, I'm just noticing this now).

Reply:
Fixed ConstructHeadersPayload problem.